### PR TITLE
Optimize import time for `conda activate`

### PR DESCRIPTION
**Pull request by mbargull**
### Description

Avoid some imports to speed up `conda activate`.

Fixes gh-13567.

| `python`<br />`=3.8.18`,<br />`ruamel.yaml`<br />`=0.16.12` | `python`<br />`=3.11.7`,<br />`ruamel.yaml`<br />`=0.16.12` | `python`<br />`=3.8.18`,<br />`ruamel.yaml` <br /> `=0.18.5` | `python`<br />`=3.11.7`,<br />`ruamel.yaml`<br />`=0.18.5` | `python`<br />`=3.8.18`,<br />`ruamel.yaml` <br /> `=0.18.6` | `python`<br />`=3.11.7`,<br />`ruamel.yaml`<br />`=0.18.6` | changes |
|-------|-------|-------|-------|-------|-------|-------|
| 0.089 | 0.093 | 0.093 | 0.098 | 0.090 | 0.95 | base commit 59af01b from `main` |
| 0.084 | 0.088 | 0.089 | 0.093 | | | `Avoid costly .gateways.logging import for activate` |
| 0.086 | 0.091 | 0.091 | 0.096 | | | `Avoid costly argparse import for activate` |
| 0.086 | 0.090 | 0.090 | 0.095 | | | `Use packaging.version in .deprecations only if needed` |
| 0.080 | 0.081 | 0.086 | 0.093 | | | `Avoid costly inspect.getmodule in .deprecations` |
| 0.086 | 0.090 | 0.090 | 0.095 | | | `Defer platformdirs import for conda activate` |
| 0.068 | 0.071 | 0.074 | 0.081 | 0.069 | 0.072 | everything above combined |
| 0.065 | 0.068 | | | 0.065 | 0.68 | everything above plus `Defer tempfile import for auxlib.compat` |

(Edit: Added `ruamel.yaml=0.18.6` timings after improvements re. https://sourceforge.net/p/ruamel-yaml/tickets/504/ .)

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- ~~[ ] Add / update necessary tests?~~
- ~~[ ] Add / update outdated documentation?~~

### CODE REVIEW

File: conda/deprecations.py [hunk 1]
```diff
@@ -285,6 +335,20 @@ def _get_module(self, stack: int) -> tuple[ModuleType, str]:
 
         try:
             frame = sys._getframe(2 + stack)
+            # Shortcut finding the module by manually inspecting loaded modules.
+            if hasattr(frame, "f_code") and hasattr(frame.f_code, "co_filename"):
```
Comment (kenodegard) [inline conda/deprecations.py:339]: duck typing instead?

and looking at this again it appears that 338-354 should actually be defined outside this try-except:

```python
try:
    frame = sys._getframe(2 + stack)
except IndexError:
    # IndexError: 2 + stack is out of range
    raise DeprecatedError("unable to determine the calling module")

try:
    filename = frame.f_code.co_filename
except AttributeError:
    # AttributeError: frame.f_code.co_filename is undefined
    pass
else:
    for module in sys.modules.values():
        ...

import inspect
...
```

File: conda/deprecations.py [hunk 2]
```diff
@@ -285,6 +335,20 @@ def _get_module(self, stack: int) -> tuple[ModuleType, str]:
```
Comment (kenodegard) [inline conda/deprecations.py:335]: FYI the original `import inspect` is still on the line above

Comment (mbargull) [reply conda/deprecations.py:336]: Argh, forgot to add that to the commit!

File: conda/deprecations.py (hunk 1 above)
Comment (mbargull) [reply conda/deprecations.py]: > duck typing instead?

I don't mind either way; added your suggestion!

> and looking at this again it appears that 338-354 should actually be defined outside this try-except:

Ah, I didn't even look at the exception type. Makes sense -- added!

File: conda/deprecations.py [hunk 3]
```diff
@@ -22,17 +25,64 @@ class DeprecatedError(RuntimeError):
...
+
+        if self._version_object is None:
+            try:
+                self._version_object = parse(self._version_str)
+            except TypeError:
+                self._version_object = parse("0.0.0.dev0+placeholder")
+        return self._version_object
+
+    @staticmethod
+    def _get_version_tuple(version_str: str) -> tuple[int]:
```
[the remark this answers is not in the document]
Comment (mbargull) [reply conda/deprecations.py]: I always forget the ellipsis :facepalm: -- thanks, added!

**Comment by mbargull** [comment]
Thanks for the early review, Ken!
I'll do some timing on this and possibly look for more pieces (but I think upstream work in `ruamel.yaml` would cover most remaining stuff already).

**Comment by kenodegard** [comment]
@mbargull the benchmarking tests (via CodSpeed) don't run correctly ATM but can we include a benchmark test (mark it with `@pytest.mark.benchmark`) so we can track this better once we get CodSpeed working?

**Comment by mbargull** [comment]
> @mbargull the benchmarking tests (via CodSpeed) don't run correctly ATM but can we include a benchmark test (mark it with `@pytest.mark.benchmark`) so we can track this better once we get CodSpeed working?

As for really benchmarking, I don't know, actually.
These activation/import times are very sensitive, so you'd need to be careful of the setup and how you'd compare results.
You need a reasonably quiet system (e.g., not too much CPU, disk activity in the background) and definitely should run timings for the merge base in the same job to get valid target data points to compare against.
The Python interpreter gives us a rather small budget to work with and is at the same time very sensitive to background work (there are just many CPU cycles and disk accesses involved in these startup/import phases...).
I have no idea how CodSpeed works/operates and as such can't really tell how things would work out.

We could instead rather have some proxies we could use to assess this, though, by not looking at the timings themselves but rather look for "known bad/slow imports".
E.g., parse the output of `PYTHONPROFILEIMPORTTIME=1 conda shell.${SHELL_NAME} activate 2>&1 > /dev/null` and ensure that `inspect` etc. is not imported.
(This would be platform and shell specific, e.g., `tempfile` would probably be imported for `shell.cmd.exe`.)

I wanted to open an issue for us to track addition of basic performance regression tests anyway, so we can avoid regressing on this, but also the massive memory+cpu one from gh-13541 .

**Comment by dholth** [comment]
`len(sys.modules)` would be a pretty good proxy.

File: conda/deprecations.py [hunk 4]
```diff
@@ -22,17 +25,65 @@ class DeprecatedError(RuntimeError):
 # inspired by deprecation (https://deprecation.readthedocs.io/en/latest/) and
 # CPython's warnings._deprecated
 class DeprecationHandler:
-    _version: Version
+    _version_str: str
+    _version_tuple: tuple[int, ...]
```
Comment (kenodegard) [inline conda/deprecations.py:29]:
```suggestion
    _version_str: str | None
    _version_tuple: tuple[int, ...] | None
```

File: conda/deprecations.py [hunk 5]
```diff
@@ -22,17 +25,65 @@ class DeprecatedError(RuntimeError):
...
         """Factory to create a deprecation handle for the specified version.
 
         :param version: The version to compare against when checking deprecation statuses.
         """
+        self._version_str = None
+        self._version_object = None
+        self._version_tuple = None
+        if version is not None:
+            self._version_str = str(version)
+            if not isinstance(version, str):
```
Comment (kenodegard) [inline conda/deprecations.py:42]: Technically anything could be passed in, so just checking for `str` is incomplete?

```suggestion
            if isinstance(version, Version):
```

**Review by kenodegard — APPROVED** [review]
Two minor suggestions

File: conda/deprecations.py (hunk 5 above)
Comment (mbargull) [reply conda/deprecations.py:42]:
```suggestion
            if not isinstance(version, str):
                from packaging.version import Version
                
                if isinstance(version, Version):
```
otherwise we'd need to import `packaging.version` unconditionally again.

Comment (kenodegard) [reply conda/deprecations.py:42]: actually why don't we just change the version input to only expect `str`? we initialize the `DeprecationHandler` as:

```python
deprecated = DeprecationHandler(__version__)
```

and I wouldn't expect anyone to pass in a `Version` object (even in our tests we only test with passing in `None` or `str`):

```python
deprecated = DeprecationHandler(Version(...))
```

Comment (mbargull) [reply conda/deprecations.py:42]: Yes, I'd be in favor of that.
I just kept the `Version` input because it was there and I didn't care to look if/how things are actually used :D.

Comment (mbargull) [reply conda/deprecations.py]: Done and simplified the code accordingly; looks clearer with much less boilerplate.

**Review by kenodegard — DISMISSED** [review]

**Review by kenodegard — APPROVED** [review]

**Comment by jezdez** [comment]
For the record this change should never have been merged since it implements a custom version parsing scheme, which is know to be a source of hard-debug issues in packaging ecosystems.

**Comment by mbargull** [comment]
> For the record this change should never have been merged since it implements a custom version parsing scheme, which is known to be a source of hard-debug issues in packaging ecosystems.

Explain? In case you are referring to `.deprecations._get_version_tuple`/`.deprecations._version_less_than`:
These are entirely internal functions and `.deprecations.deprecated` *should* only be used with `conda`'s own version as input.
And even then: The parsing therein is only used when both versions of the comparison can be converted via `tuple(int(part) for part in version.strip().split("."))` without raising any errors -- if there are any errors, it gracefully falls back to `packaging.version.parse`.

To summarize: It is not meant for general usage but even then gracefully falls back to established version parsing code.
Hence, I do not share your concerns at all.

(Rather than reverting this back to always using `packaging.version`, I'd even suggest changing the interface such that it doesn't take a string that has to be parsed expensively, if you'd really want to amend this.)

**Comment by jezdez** [comment]
> To summarize: It is not meant for general usage but even then gracefully falls back to established version parsing code. Hence, I do not share your concerns at all.

I think you've misunderstood the goal of the deprecations framework, it's supposed to be used by other conda related libraries, in support of plugin based conda ecosystem, and not just internally.

**Comment by mbargull** [comment]
> I think you've misunderstood the goal of the deprecations framework, it's supposed to be used by other conda related libraries, in support of plugin based conda ecosystem, and not just internally.

Ah, okay, I was not aware of that.
If I had been, I'd certainly been more stringent about this, i.e., included the changes from gh-13681 from the get go.
